### PR TITLE
Adjust tests to be executed in parallel

### LINKED ISSUE #1150: Parallel test execution
AT:
- every single test method (Java) in the Zeebe repository can be executed in parallel if unlimited resources are available

Follow up issue of #943

### PR DESCRIPTION
**Pull request by menski**
- add a maven profile `parallel-tests` to use forks during surefire execution
- add utility to find next free port
- adjust all test suites to use dynamic instead of static ports
- run test suite in parallel on Jenkins

Current observations indicate that the test during is shrinking from 14min to 6min. 

closes #1150

### CODE REVIEW

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java [hunk 1]
```diff
@@ -1,132 +0,0 @@
...
- */
-package io.zeebe.broker.it.clustering;
-
-import static io.zeebe.broker.it.clustering.ClusteringRule.BROKER_1_CLIENT_ADDRESS;
-import static io.zeebe.broker.it.clustering.ClusteringRule.BROKER_1_TOML;
-import static io.zeebe.broker.it.clustering.ClusteringRule.BROKER_2_CLIENT_ADDRESS;
-import static io.zeebe.broker.it.clustering.ClusteringRule.BROKER_2_TOML;
-import static io.zeebe.broker.it.clustering.ClusteringRule.BROKER_4_CLIENT_ADDRESS;
-import static io.zeebe.broker.it.clustering.ClusteringRule.BROKER_4_TOML;
-import static io.zeebe.protocol.Protocol.DEFAULT_TOPIC;
```
Comment (Zelldon) [inline qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java]: Why this test was deleted? I can remember that gossip joining with different nodes does not work in the past :sweat_smile:

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java [hunk 2]
```diff
@@ -288,39 +285,65 @@ private Topic waitForTopicAvailability(final String topicName) {
...
+          this.getClass().getClassLoader().getResourceAsStream(brokerConfigFiles[brokerId]);
+      final BrokerCfg brokerCfg = TomlConfigurationReader.read(config);
+      EmbeddedBrokerRule.assignSocketAddresses(brokerCfg);
+
+      if (brokerCfgs[0] != null) {
+        brokerCfg
+            .getCluster()
+            .setInitialContactPoints(
+                new String[] {
+                  brokerCfgs[0].getNetwork().getManagement().toSocketAddress().toString()
```
Comment (Zelldon) [inline qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java]: So they will join all the first node right?

File: broker-core/src/main/java/io/zeebe/broker/transport/GatewayComponent.java [hunk 3]
```diff
@@ -27,11 +29,14 @@
...
   public void init(final SystemContext context) {
-    final SocketBindingGatewayCfg gatewayCfg =
-        context.getBrokerConfiguration().getNetwork().getGateway();
+    final NetworkCfg network = context.getBrokerConfiguration().getNetwork();
+    final SocketBindingGatewayCfg gatewayCfg = network.getGateway();
     if (gatewayCfg.isEnabled()) {
       try {
+        final SocketBindingClientApiCfg clientApiCfg = network.getClient();
         final Gateway gateway = new Gateway(gatewayCfg.getHost(), gatewayCfg.getPort());
+        // TODO(menski): help this is horrible
```
Comment (Zelldon) [inline broker-core/src/main/java/io/zeebe/broker/transport/GatewayComponent.java]: Do you need help? :sweat_smile:

File: transport/src/main/java/io/zeebe/transport/impl/util/SocketUtil.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    public boolean hasNext() {
+      return true;
+    }
+
+    @Override
+    public SocketAddress next() {
+      return new SocketAddress(host, nextPort());
+    }
+
+    private int nextPort() {
```
Comment (Zelldon) [inline transport/src/main/java/io/zeebe/transport/impl/util/SocketUtil.java]: In general it looks fine, but one question I have.
Do we need to care about thread safety? How does this surefire parallel test execution work? Is each TestClass one JVM/Process?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

File: parent/pom.xml [hunk 5]
```diff
@@ -571,5 +571,24 @@
...
+        <plugins>
+          <plugin>
+            <groupId>org.apache.maven.plugins</groupId>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <configuration>
+              <forkCount>1C</forkCount>
+              <reuseForks>true</reuseForks>
+              <systemPropertyVariables>
+                <!-- without the leading 0 the property was not expanded -->
+                <testForkNumber>0${surefire.forkNumber}</testForkNumber>
```
Comment (Zelldon) [inline parent/pom.xml]: I'm not sure if I get this comment :see_no_evil:

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java (hunk 1 above)
Comment (menski) [reply qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java]: I didn't come up with a good solution to configure a different initial contact point while using dynamic port assignments. And I hope that the gossip test already covers the issues you mentioned https://github.com/zeebe-io/zeebe/blob/1150-parallel-test-execution/gossip/src/test/java/io/zeebe/gossip/GossipJoinTest.java#L190.

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java (hunk 2 above)
Comment (menski) [reply qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java]: Yes with this change every node other then the first will always join the first node as it was before configured in the configuration files.

File: broker-core/src/main/java/io/zeebe/broker/transport/GatewayComponent.java (hunk 3 above)
Comment (menski) [reply broker-core/src/main/java/io/zeebe/broker/transport/GatewayComponent.java]: I think we will address this in the gateway configuration issue #1135

File: transport/src/main/java/io/zeebe/transport/impl/util/SocketUtil.java (hunk 4 above)
Comment (menski) [reply transport/src/main/java/io/zeebe/transport/impl/util/SocketUtil.java]: As far as I understood the documentation if only forks are used every test class will be handed to a single fork. We currently only use a single thread in each fork so no test methods are executed in parallel. So to answer your question I think at the moment we don't need to care about thread safety.

> When using reuseForks=true and a forkCount value larger than one, test classes are handed over to the forked process one-by-one.

https://maven.apache.org/surefire/maven-surefire-plugin/examples/fork-options-and-parallel-execution.html

Comment (Zelldon) [reply transport/src/main/java/io/zeebe/transport/impl/util/SocketUtil.java]: Ok I assumed this as well. Thanks :+1:

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java (hunk 1 above)
Comment (Zelldon) [reply qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java]: Perhaps choosing, in the ClusteringRule, randomly the initial points or brokerId -1. So Node 1 connects to 0 and Node 2 to 1.

File: parent/pom.xml (hunk 5 above)
Comment (menski) [reply parent/pom.xml]: I had the issue that when I only had the surefire property inside the system property, i.e. `<testForkNumber>${surefire.forkNumber}</testForkNumber>`, that then the property `testForkNumber` was empty on test execution. I didn't understood why but also didn't investigated further. So as a workaround I added a `0` as prefix as then the property is set and still parseable as integer.

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java (hunk 1 above)
Comment (menski) [reply qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/GossipDifferentNodeJoinTest.java]: > So Node 1 connects to 0 and Node 2 to 1.

I can do that if you want but then we still don't need the test case cause it is identical with `GossipClusteringTest.java` if I remember correctly.

File: parent/pom.xml (hunk 5 above)
Comment (Zelldon) [reply parent/pom.xml]: ok weird :sweat_smile: http://maven.40175.n5.nabble.com/SureFire-2-16-doesn-t-respect-forkCount-tp5771850p5771854.html

**Review by ChrisKujawa — APPROVED** [review]

File: parent/pom.xml (hunk 5 above)
Comment (menski) [reply parent/pom.xml]: Oh thanks for this I will then change it to `$${` instead of `0${` as this makes more sense :+1:

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 0f7c87c799d6f39110071f3db5e0e02cffd83a12
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md

---
This message was auto-generated by https://gitcop.com

**Comment by menski** [comment]
It seems that currently sometimes broker survive the end of a test case and continue to run, which in the end leads to an exhausting of the available port range for a test runner fork. I have to investigate further why this is happening before we can merge the changes.

**Comment by menski** [comment]
It seems that one problem was that the container run out of memory (8GB). We will now run the test suite with a 12GB container.

**Comment by menski** [comment]
I made the following changes since the review hints:
- use `$${surefire.forkNumber}` in parent pom to prevent maven from expanding the variable
- the `ClusteringRule` in the broker IT suite will configure cluster members to join the previous node instead of always joining the first node
- the `SocketUtil` now owns a port range of 100 ports to prevent exhausting by not freed ports
- the Zeebe Java test suite is now run a container with 12GB memory reserved

bors r+

**Pull request merged**